0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+
+- Ephemeral secrets (for generation and replication of ephemeral secrets)
+
+- CPU change (for detecting wether the CPU has changed)
+
+- CHURP (secret sharing scheme)
+
+- Key derivation center (secret sharing scheme)
+
+### App trait
+
+Every application should implement the following trait.
+
+```rust
+/// Key manager application.
+pub trait App {
+  /// Register RPC methods to the dispatcher.
+  ///
+  /// Registered methods are exposed via enclave RPC to the local host
+  /// or remote clients.
+  fn register(&self, ...);
+
+  /// Initialize the application on startup.
+  ///
+  /// Use runtime ID, untrusted local storage, host data, consensus verifier,
+  /// runtime identity, etc. to start the application.
+  fn init(&self, ...);
+}
+```
+
+Each application should register RPC methods, prefixed with the application
+name.
+
+#### Example 1
+
+```rust
+/// Master secrets key manager application.
+pub trait MasterSecrets {
+  fn generate(&self);
+  fn load(&self);
+  fn replicate(&self);
+  fn key_pair(&self);
+  fn private_key(&self);
+  fn public_key(&self);
+  fn symmetric_key(&self);
+  fn update_status(&self);
+}
+```
+
+Methods:
+
+- `master-secrets/generate`
+
+- `master-secrets/load`
+
+- `master-secrets/replicate`
+
+- `master-secrets/key_pair`
+
+- `master-secrets/private_key`
+
+- `master-secrets/public_key`
+
+- `master-secrets/symmetric_key`
+
+- `master-secrets/update_status`
+
+#### Example 2
+
+```rust
+/// CPU change detection key manager application.
+pub trait CPUChangeDetection {
+  fn encrypt(&self);
+  fn decrypt(&self);
+}
+```
+
+Methods:
+
+- `cpu-change/encrypt`
+
+- `cpu-change/decrypt`
+
+### App worker
+
+Each key manager application should have a dedicated worker on the host node
+responsible for communicating with the app and ensuring its consensus view
+is up-to-date.
+
+```go
+// AppWorker is a key manager application worker.
+type AppWorker interface {
+  // Start starts the worker.
+  Start()
+  // Stop stops the worker.
+  Stop()
+}
+```
+
+#### Example
+
+A master secret worker should be responsible for participating in the master
+secret protocol.
+
+## CHURP
+
+CHURP is a proactive secret sharing scheme in which the committee of nodes
+storing a secret can change over time.
+
+### Requirements
+
+- Support multiple secrets.
+
+  - Each group of clients can share a unique secret.
+
+  - Secrets may vary in terms of committee size, handoff intervals,
+    security levels, etc.
+
+  - State re-encryption should be possible with keys derived from
+    a new shared secret.
+
+- Key manager restarts should not lose shares or committed data.
+
+### Protocol
+
+#### Offline
+
+- Assign non-zero ID numbers to all nodes and associate them with their
+  public keys.
+
+- The key manager owner selects a Schnorr group based on security
+  requirements and prepares the CHURP configuration and access policy.
+
+  - The Schnorr group can also be generated by one of the enclaves.
```

Comment:
What are the tradeoffs?

Reply:
In theory, the key manager owner could find a prime with security weaknesses and use that one. So in general, it is always better to use randomly selected primes, curves, etc. or have a proof that selection was random. 
Since this is done only once, the prime generation complexity is not a problem. Just extra step in the protocol.